packages/backend/src/open_payments/grant/service.ts

```diff
@@ -89,6 +101,140 @@ async function updateGrant(
     .withGraphFetched('authServer')
 }
 
+async function getOrCreateGrant(
+  deps: ServiceDependencies,
+  options: GrantOptions
+): Promise<Grant | GrantError> {
+  const existingGrant = await getExistingGrant(deps, options)
+
+  if (!existingGrant) {
+    return requestNewGrant(deps, options)
+  }
+
+  if (existingGrant.expired) {
+    return (
+      (await rotateTokenAndUpdateGrant(deps, existingGrant)) ??
+      (await requestNewGrant(deps, options))
+    )
+  }
+
+  return existingGrant
+}
+
+export async function getExistingGrant(
+  deps: ServiceDependencies,
+  options: GrantOptions
+): Promise<Grant | undefined> {
+  return await Grant.query(deps.knex)
+    .findOne({
+      accessType: options.accessType
+    })
+    .whereNull('deletedAt')
+    .andWhere('authServer.url', options.authServer)
+    .andWhere('accessActions', '@>', options.accessActions) // all options.accessActions are a subset of saved accessActions
```

Comment:
e.g. if we have `[AccessActions.Create, AccessActions.Complete, AccessActions.Read]` and we are trying to getOrCreate a grant with `[AccessActions.Create]` we will properly return it, but if one of the elements in the requested list doesnt exist, we will not be able to fetch it

Reply:
It looks like this differs slightly from the `getGrant` which finds one that has strictly equal `accessActions`.  Its not quite clear what the ramifications of this will be. Can you speak to this difference?

Reply:
before, if we already had a grant for `[Read, Create]` and then we were trying to getOrCreate a grant with `[Read]`, it would fail to get the existing one, even though the existing grant has the sufficient access requirements

Reply:
I suppose we're getting rid of the `getGrant` here then (in subsequent PR)? looks like its unused but if we keep it for some reason I guess we should update to filter out by `deletedAt` (and maybe this same `accessActions` where clause). But I imagine we can just remove it.

Reply:
Yes exactly, will be removing the unused methods